# DNAmDeconv pipeline: Citations

## DNAmDeconv
>Add citation after paper is published.

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)
> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)
> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline Tools
* <u>[meth_atlas](https://www.nature.com/articles/s41467-018-07466-6)</u>
    > Moss, J.; Magenheim, J.; Neiman, D.; Zemmour, H.; Loyfer, N.; Korach, A.; Samet, Y.; Maoz, M.; Druid, H.; Arner, P.; et al. Comprehensive human cell-type methylation atlas reveals origins of circulating cell-free DNA in health and disease. Nat. Commun. 2018, 9, 5068.
* <u>[CIBERSORT](https://www.nature.com/articles/nmeth.3337)</u>
    > Newman, A., Liu, C., Green, M. et al. Robust enumeration of cell subsets from tissue expression profiles. Nat Methods 12, 453–457 (2015).
* <u>[EpiDISH](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1511-5)</u>
    >Teschendorff, A.E., Breeze, C.E., Zheng, S.C. et al. A comparison of reference-based algorithms for correcting cell-type heterogeneity in Epigenome-Wide Association Studies. BMC Bioinformatics 18, 105 (2017)
* <u>[EpiSCORE](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02126-9)</u>
    >Teschendorff, A.E., Zhu, T., Breeze, C.E. et al. EPISCORE: cell type deconvolution of bulk tissue DNA methylomes from single-cell RNA-Seq data. Genome Biol 21, 221 (2020)
* <u>[Methyl_Resolver](https://www.nature.com/articles/s42003-020-01146-2)</u>
    >Arneson, D., Yang, X. & Wang, K. MethylResolver—a method for deconvoluting bulk DNA methylation profiles into known and unknown cell contents. Commun Biol 3, 422 (2020)
* <u>[PRMeth](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-022-04893-7)</u>
    >He, D., Chen, M., Wang, W. et al. Deconvolution of tumor composition using partially available DNA methylation data. BMC Bioinformatics 23, 355 (2022)
* <u>[MeDeCom](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1182-6)</u>
    >Lutsik, P., Slawski, M., Gasparoni, G. et al. MeDeCom: discovery and quantification of latent components of heterogeneous methylomes. Genome Biol 18, 55 (2017)
* <u>[CelFiE](https://www.nature.com/articles/s41467-021-22901-x)</u>
    >Caggiano, C., Celona, B., Garton, F. et al. Comprehensive cell type decomposition of circulating cell-free DNA with CelFiE. Nat Commun 12, 2717 (2021)
* <u>[limma](https://academic.oup.com/nar/article/43/7/e47/2414268)</u>
    >Matthew E. Ritchie, Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, Gordon K. Smyth, limma powers differential expression analyses for RNA-sequencing and microarray studies, Nucleic Acids Research, Volume 43, Issue 7, 20 April 2015
* <u>[BEDTools](https://academic.oup.com/bioinformatics/article/26/6/841/244688)</u>
    >Aaron R. Quinlan, Ira M. Hall, BEDTools: a flexible suite of utilities for comparing genomic features, Bioinformatics, Volume 26, Issue 6, March 2010, Pages 841–842
* <u>[DSS](https://academic.oup.com/bioinformatics/article/32/10/1446/1743267?login=false)</u>
    >Yongseok Park, Hao Wu, Differential methylation analysis for BS-seq data under general experimental design, Bioinformatics, Volume 32, Issue 10, May 2016, Pages 1446–1453
## Software packaging/containerisation tools
* <u>[Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)</u>
    >Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2
* <u>[Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)</u>
    >Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459
